lib/util/Ranges.v3

```diff
@@ -51,31 +51,29 @@ component Ranges {
 	}
 	// Sort the range {r} in place using Quicksort using the {cmp} function.
 	def quicksort<T>(r: Range<T>, cmp: (T, T) -> bool) {
-		if (r.length <= 1) return;
-		if (r.length == 2) {
-			var a = r[0], b = r[1];
-			if (cmp(b, a)) { r[0] = b; r[1] = a; }
-			return;
-		}
-		def pi = partition(r, cmp);
-		if (pi > 1) quicksort(r[0 ... pi], cmp);
-		if (pi < r.length) quicksort(r[pi ...], cmp);
-	}
-	private def partition<T>(r: Range<T>, cmp: (T, T) -> bool) -> int {
-		var i = 0, j = r.length - 1;
-		def pivot = r[j];
-		while (i <= j) {
-			var a = r[i];
-			if (cmp(a, pivot)) {
-				i++;
-			} else {
-				r[i] = r[j];
-				r[j] = a;
-				j--;
-			}
-		}
-		return i;
-	}
+    if (r.length <= 1) return;
+    if (r.length == 2) {
+      var a = r[0], b = r[1];
+      if (cmp(b, a)) { r[0] = b; r[1] = a; }
+      return;
+    }
+    def pi = partition(r, cmp);
+    if (pi > 1) quicksort(r[0 ... pi], cmp);
+    if (pi < r.length - 2) quicksort(r[pi + 1 ...], cmp);
+  }
+  private def partition<T>(r: Range<T>, cmp: (T, T) -> bool) -> int {
+    def pivot = r[r.length - 1];
+    var i = 0;
+    for (j = 0; j < r.length; j++) {
+      if (cmp(r[j], pivot) || j == r.length - 1) {
```

Comment:
Is it possible that cmp(r[j], pivot) is never true, so only the last iteration will increment i to 1, and then the return at the end will return i-1 == 0?

Can you write a test that tests all possible 4-length arrays containing the elements 1, 2, 3, 4? E.g. [1, 1, 1, 1], [1, 1, 1, 2] etc? I think that will be guaranteed to hit all possible cases of partition.

Reply:
I think the case you mentioned still works, since if `cmp(r[j], pivot)` is never true, then `pivot` is less than all other `r[j]` and thus should be at index `0` when the function returns.